tests/test_dse.py

```diff
@@ -1787,6 +1787,67 @@ def test_blocking_options(self, rotate):
         assert np.isclose(expected, norm(u1), rtol=1e-5)
         assert np.isclose(expected, norm(u2), rtol=1e-5)
 
+    def test_ftemps_option(self):
+        """
+        Test the compiler option `cire-ftemps=True`. This will make CIRE use
+        TempFunctions, rather than Arrays, to create temporaries, thus giving
+        control over allocation and deallocation to the user.
+        """
+        grid = Grid(shape=(30, 30, 30))
+        x, y, z = grid.dimensions
+        t = grid.stepping_dim
+
+        nthreads = 2
+        x0_blk0_size = 8
+        y0_blk0_size = 8
+
+        f = Function(name='f', grid=grid)
+        u = TimeFunction(name='u', grid=grid, space_order=3)
+        u1 = TimeFunction(name="u", grid=grid, space_order=3)
+        u2 = TimeFunction(name="u", grid=grid, space_order=3)
+
+        f.data_with_halo[:] = 1.
+        u.data_with_halo[:] = 0.32
+        u1.data_with_halo[:] = 0.32
+        u2.data_with_halo[:] = 0.32
+
+        eqn = Eq(u.forward, ((u[t, x, y, z] + u[t, x+1, y+1, z+1])*3*f +
+                             (u[t, x+2, y+2, z+2] + u[t, x+3, y+3, z+3])*3*f + 1))
+
+        op0 = Operator(eqn, opt=('noop', {'openmp': True}))
+        op1 = Operator(eqn, opt=('advanced', {'openmp': True, 'cire-mincost-sops': 1,
+                                              'cire-ftemps': True}))
+        op2 = Operator(eqn, opt=('advanced-fsg', {'openmp': True, 'cire-mincost-sops': 1,
+                                                  'cire-ftemps': True}))
+
+        op0(time_M=1, nthreads=nthreads)
+
+        # TempFunctions expect an override
+        try:
+            op1(time_M=1, u=u1)
+        except InvalidArgument:
+            assert True
+        except:
+            assert False
+
+        # Prepare to run op1
+        shape = [nthreads, x0_blk0_size, y0_blk0_size, grid.shape[-1]]
```

Comment:
ohhh...  I had not thought about this. How is one supposed to know what to pass to `make`, here - if it's not always the `grid.shape`?

This seems to mean that the caller has to be aware of the details of the generated code. I guess we can live with that, but I'd prefer more safeguards if possible.  Could there potentially be some protection in `arg_values` that ensures the passed temporary is sufficiently large, given the rest of the arguments?

The sequence of events I'm worried about is:
 * devito version X produces C code that needs temporaries of a certain shape
 * application hard-codes temporaries of that shape using a construction very similar to this line
 * devito version X+1 produces C code that needs a slightly larger temporary, but code still passes same size
 * segfaults, incorrect results, chaos reigns
 
Alternatively, or perhaps as well, could the arguments to `make` be _all_ the arguments that will be passed to the operator, and make can extract whatever values it needs from those?

Reply:
>This seems to mean that the caller has to be aware of the details of the generated code.

Well, the idea was that you query the TempFunction's shape. This gives you the _symbolic_ shape of the TempFunction from which you can infer what's required.

> Alternatively, or perhaps as well, could the arguments to make be all the arguments that will be passed to the operator, and make can extract whatever values it needs from those?

yes, if this is feasible on your side, it should be straightforward on our side. So the idea would be (sort of):

```
def make(..., **kwargs):
    # Sanity checks...
  
    try:  
        shape = tuple(i.subs(kwargs) for I in self.shape)
    except ...:
        raise ValueError('Don't know how to derive shape for ...') 
```

?

Reply:
Ah, yeah I hadn't appreciated that we could get the symbolic shape. That resolves my concern for now, I think.